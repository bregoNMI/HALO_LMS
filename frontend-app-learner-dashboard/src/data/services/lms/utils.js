import queryString from 'query-string';
import { getAuthenticatedHttpClient } from '@edx/frontend-platform/auth';

/**
 * get(url)
 * simple wrapper providing an authenticated Http client get action
 * @param {string} url - target url
 */
export const get = (...args) => getAuthenticatedHttpClient().get(...args);
/**
 * post(url, data)
 * simple wrapper providing an authenticated Http client post action
 * queryString.stringify is used to convert the object to query string with = and &
 * @param {string} url - target url
 * @param {object|string} body - post payload
 */
export const post = (url, body) => getAuthenticatedHttpClient().post(url, queryString.stringify(body));

export const client = getAuthenticatedHttpClient;

/**
 * stringifyUrl(url, query)
 * simple wrapper around queryString.stringifyUrl that sets skip behavior
 * @param {string} url - base url string
 * @param {object} query - query parameters
 */
export const stringifyUrl = (url, query) => queryString.stringifyUrl(
  { url, query },
  { skipNull: true, skipEmptyString: true },
);
