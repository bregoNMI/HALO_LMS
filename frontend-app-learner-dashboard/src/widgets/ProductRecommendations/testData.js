export const getCoursesWithType = (courseTypes, parameters = true) => {
  const courses = [];
  const marketingUrl = parameters
    ? 'https://www.edx.org/course/some-course?utm_source=source'
    : 'https://www.edx.org/course/some-course';

  courseTypes.forEach((type) => {
    courses.push({
      title: 'Introduction to Computer Science',
      courseRunKey: 'course-v1:Test+Course+2022T2',
      marketingUrl,
      courseType: type,
      image: {
        src: 'https://www.image-2.com/ed79a49b-64c1-48d2-afdc-054bf921e38d-6a76ceb47dea.small.jpg',
      },
      owners: [
        {
          key: 'HarvardX',
          name: 'Harvard University',
          logoImageUrl: 'http://www.image.com/ef72daf3-c9a1-4c00-ba37-b3514392bdcf-8839c516815a.png',
        },
      ],
    });
  });

  return courses;
};

export const mockFooterRecommendationsHook = {
  default: { isExperimentActive: false, inRecommendationsVariant: true },
  activeControl: { isExperimentActive: true, inRecommendationsVariant: false },
  activeTreatment: { isExperimentActive: true, inRecommendationsVariant: true },
};

export const mockCrossProductCourses = getCoursesWithType(['executive-education-2u', 'bootcamp-2u']);
export const mockOpenCourses = getCoursesWithType(['verified-audit', 'audit', 'verified', 'course']);
export const mockFallbackOpenCourse = getCoursesWithType(['course'], false);

export const mockCrossProductResponse = {
  crossProductCourses: mockCrossProductCourses,
  amplitudeCourses: mockOpenCourses,
};

export const mockAmplitudeResponse = {
  amplitudeCourses: mockOpenCourses,
};
