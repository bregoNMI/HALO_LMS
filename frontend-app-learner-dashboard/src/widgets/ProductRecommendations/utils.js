import PropTypes from 'prop-types';
import { executiveEducation, course, bootCamp } from './constants';

export const courseShape = {
  uuid: PropTypes.string,
  title: PropTypes.string,
  image: PropTypes.shape({
    src: PropTypes.string,
  }),
  prospectusPath: PropTypes.string,
  owners: PropTypes.arrayOf(
    PropTypes.shape({
      key: PropTypes.string,
      name: PropTypes.string,
      logoImageUrl: PropTypes.string,
    }),
  ),
  activeCourseRun: PropTypes.shape({
    key: PropTypes.string,
    marketingUrl: PropTypes.string,
  }),
  courseType: PropTypes.string,
};

export const courseTypeToProductTypeMap = {
  course,
  'verified-audit': course,
  verified: course,
  audit: course,
  'credit-verified-audit': course,
  'spoc-verified-audit': course,
  professional: 'Professional Certificate',
  'bootcamp-2u': bootCamp,
  'executive-education-2u': executiveEducation,
  'executive-education': executiveEducation,
  masters: "Master's",
  'masters-verified-audit': "Master's",
};

export const courseTypeToProductLineMap = {
  [executiveEducation]: 'executive-education',
  [bootCamp]: 'boot-camps',
  [course]: 'open-courses',
};

export const convertCourseRunKeyToCourseKey = (courseRunKey) => {
  const newKeyFormat = courseRunKey.includes('+');
  if (newKeyFormat) {
    const splitCourseRunKey = courseRunKey.split(':').slice(-1)[0];
    const splitCourseKey = splitCourseRunKey.split('+').slice(0, 2);
    return `${splitCourseKey[0]}+${splitCourseKey[1]}`;
  }
  const splitCourseKey = courseRunKey.split('/').slice(0, 2);
  return `${splitCourseKey[0]}+${splitCourseKey[1]}`;
};

export const wait = (time) => new Promise((resolve) => {
  setTimeout(resolve, time);
});
