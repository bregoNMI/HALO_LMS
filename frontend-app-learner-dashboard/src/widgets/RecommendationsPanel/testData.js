// eslint-disable-next-line
export default {"courses":[{"courseKey":"ETSx+TOEFLx","logoImageUrl":"https://prod-discovery.edx-cdn.org/organization/logos/9d9e1a30-c34d-4ad1-8c5a-d2410db8c123-8beea336c2a4.png","marketingUrl":"https://www.edx.org/course/toefl-test-preparation-the-insiders-guide?utm_source=lms_catalog_service_user&utm_medium=affiliate_partner","title":"TOEFL® Test Preparation: The Insider’s Guide"},{"courseKey":"HarvardX+CS50P","logoImageUrl":"https://prod-discovery.edx-cdn.org/organization/logos/44022f13-20df-4666-9111-cede3e5dc5b6-2cc39992c67a.png","marketingUrl":"https://www.edx.org/course/cs50s-introduction-to-programming-with-python?utm_source=lms_catalog_service_user&utm_medium=affiliate_partner","title":"CS50's Introduction to Programming with Python"},{"courseKey":"UQx+IELTSx","logoImageUrl":"https://prod-discovery.edx-cdn.org/organization/logos/8554749f-b920-4d7f-8986-af6bb95290aa-f336c6a2ca11.png","marketingUrl":"https://www.edx.org/course/ielts-academic-test-preparation?utm_source=lms_catalog_service_user&utm_medium=affiliate_partner","title":"IELTS Academic Test Preparation"},{"courseKey":"QUx+QU01X02","logoImageUrl":"https://prod-discovery.edx-cdn.org/organization/logos/7b6ca3d5-1030-408f-b1a4-fb140db1854e-7337e1e6deaf.png","marketingUrl":"https://www.edx.org/course/arabic-for-non-arabic-speakers?utm_source=lms_catalog_service_user&utm_medium=affiliate_partner","title":"Arabic for non-Arabic Speakers"},{"courseKey":"HarvardX+CS50W","logoImageUrl":"https://prod-discovery.edx-cdn.org/organization/logos/44022f13-20df-4666-9111-cede3e5dc5b6-2cc39992c67a.png","marketingUrl":"https://www.edx.org/course/cs50s-web-programming-with-python-and-javascript?utm_source=lms_catalog_service_user&utm_medium=affiliate_partner","title":"CS50's Web Programming with Python and JavaScript"}],"isControl":true};
