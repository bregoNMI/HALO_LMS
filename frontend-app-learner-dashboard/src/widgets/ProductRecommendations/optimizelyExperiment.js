import { StrictDict } from 'utils';
import optimizelyClient from './optimizely';

export const PRODUCT_RECOMMENDATIONS_EXP_KEY = 'learner_dashboard_product_recommendations_exp';
export const PRODUCT_RECOMMENDATIONS_EXP_VARIATION = 'learner_dashboard_product_recommendations_enabled';

export const eventNames = StrictDict({
  productRecommendationsViewed: 'product_recommendations_viewed',
  productHeaderClicked: 'product_header_clicked',
  productCardClicked: 'product_card_clicked',
  courseCardClicked: 'course_card_clicked',
});

export const activateProductRecommendationsExperiment = (userId, userAttributes) => {
  const variation = optimizelyClient?.activate(
    PRODUCT_RECOMMENDATIONS_EXP_KEY,
    userId,
    userAttributes,
  );

  return {
    experimentActivated: variation !== null,
    inExperimentVariant: variation === PRODUCT_RECOMMENDATIONS_EXP_VARIATION,
  };
};

export const trackProductRecommendationsViewed = (userId, userAttributes = {}) => {
  optimizelyClient.track(eventNames.productRecommendationsViewed, userId, userAttributes);
};

export const trackProductHeaderClicked = (userId, userAttributes = {}) => {
  optimizelyClient.track(eventNames.productHeaderClicked, userId, userAttributes);
};

export const trackProductCardClicked = (userId, userAttributes = {}) => {
  optimizelyClient.track(eventNames.productCardClicked, userId, userAttributes);
};

export const trackCourseCardClicked = (userId, userAttributes = {}) => {
  optimizelyClient.track(eventNames.courseCardClicked, userId, userAttributes);
};
