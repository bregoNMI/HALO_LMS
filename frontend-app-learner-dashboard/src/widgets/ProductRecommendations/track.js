import { StrictDict } from 'utils';
import { createLinkTracker, createEventTracker } from 'data/services/segment/utils';
import { courseTypeToProductLineMap, convertCourseRunKeyToCourseKey } from './utils';

export const eventNames = StrictDict({
  productCardClicked: 'edx.bi.2u-product-card.clicked',
  discoveryCardClicked: 'edx.bi.user.discovery.card.click',
  recommendationsHeaderClicked: 'edx.bi.link.recommendations.header.clicked',
  recommendationsViewed: 'edx.bi.user.recommendations.viewed',
});

export const productCardClicked = (courseRunKey, courseTitle, courseType, href) => {
  createLinkTracker(
    createEventTracker(eventNames.productCardClicked, {
      label: courseTitle,
      courserun_key: courseRunKey,
      page: 'dashboard',
      product_line: courseTypeToProductLineMap[courseType],
    }),
    href,
  );
};

export const discoveryCardClicked = (courseRunKey, courseTitle, href) => {
  createLinkTracker(
    createEventTracker(eventNames.discoveryCardClicked, {
      label: courseTitle,
      courserun_key: courseRunKey,
      page: 'dashboard',
      product_line: 'open-course',
    }),
    href,
  );
};

export const recommendationsHeaderClicked = (courseType, href) => {
  createLinkTracker(
    createEventTracker(eventNames.recommendationsHeaderClicked, {
      page: 'dashboard',
      product_line: courseTypeToProductLineMap[courseType],
    }),
    href,
  );
};

export const recommendationsViewed = (isControl, recommenderGroup, courseRunKey) => {
  createEventTracker(eventNames.recommendationsViewed, {
    is_control: isControl,
    productRecommenderGroup: recommenderGroup,
    page: 'dashboard',
    course_key: courseRunKey ? convertCourseRunKeyToCourseKey(courseRunKey) : '',
  });
};
