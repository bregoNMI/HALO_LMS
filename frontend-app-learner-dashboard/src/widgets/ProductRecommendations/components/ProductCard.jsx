import React from 'react';
import PropTypes from 'prop-types';
import {
  Badge,
  Card,
  Truncate,
  Hyperlink,
} from '@openedx/paragon';
import { getAuthenticatedUser } from '@edx/frontend-platform/auth';

import { trackProductCardClicked, trackCourseCardClicked } from '../optimizelyExperiment';
import { productCardClicked, discoveryCardClicked } from '../track';
import { bootCamp, executiveEducation } from '../constants';

const ProductCard = ({
  title,
  subtitle,
  headerImage,
  courseRunKey,
  schoolLogo,
  courseType,
  url,
}) => {
  const handleClick = (type, link) => {
    const crossProductTypes = [executiveEducation, bootCamp];
    const userId = getAuthenticatedUser().userId.toString();

    if (crossProductTypes.includes(type)) {
      trackProductCardClicked(userId);
      productCardClicked(courseRunKey, title, type, link);
    } else {
      trackCourseCardClicked(userId);
      discoveryCardClicked(courseRunKey, title, link);
    }
  };

  const getRedirectUrl = (link) => {
    const urlObj = new URL(link);
    const hasQueryStringParameters = urlObj.search !== '';

    if (hasQueryStringParameters) {
      return `${link}&linked_from=recommender`;
    }

    return `${link}?linked_from=recommender`;
  };

  const redirectUrl = getRedirectUrl(url);

  return (
    <Card
      className="base-card d-flex text-decoration-none"
      as={Hyperlink}
      destination={redirectUrl}
      onClick={() => {
        handleClick(courseType, redirectUrl);
      }}
      isClickable
    >
      <Card.ImageCap
        src={headerImage}
        srcAlt={`header image for ${title}`}
        logoSrc={schoolLogo}
        logoAlt={`logo for ${subtitle}`}
      />
      <Card.Header
        className="mt-2"
        title={(
          <Truncate lines={3} ellipsis="â€¦" className="product-card-title font-weight-bold">
            {title}
          </Truncate>
        )}
        subtitle={(
          <Truncate lines={1} className="product-card-subtitle font-weight-normal">
            {subtitle}
          </Truncate>
      )}
      />
      <Card.Section>
        <div className="product-badge position-absolute">
          <Badge className="bg-light-500 text-dark-500">{courseType}</Badge>
        </div>
      </Card.Section>
    </Card>
  );
};

ProductCard.propTypes = {
  title: PropTypes.string.isRequired,
  subtitle: PropTypes.string.isRequired,
  headerImage: PropTypes.string.isRequired,
  courseRunKey: PropTypes.string.isRequired,
  schoolLogo: PropTypes.string.isRequired,
  courseType: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
};

export default ProductCard;
