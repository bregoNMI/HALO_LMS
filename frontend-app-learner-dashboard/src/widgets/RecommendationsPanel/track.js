import { StrictDict } from 'utils';
import { createLinkTracker, createEventTracker } from 'data/services/segment/utils';
import track from 'tracking';

export const eventNames = StrictDict({
  recommendedCourseClicked: 'edx.bi.user.recommended.course.click',
});

export const linkNames = StrictDict({
  findCoursesWidget: 'learner_home_widget_explore',
});

export const findCoursesWidgetClicked = (href) => track.findCourses.findCoursesClicked(href, {
  linkName: linkNames.findCoursesWidget,
});

export const recommendedCourseClicked = (courseKey, isControl, href) => createLinkTracker(
  createEventTracker(eventNames.recommendedCourseClicked, {
    course_key: courseKey,
    is_control: isControl,
    page: 'dashboard',
  }),
  href,
);

export default {
  findCoursesWidgetClicked,
  recommendedCourseClicked,
};
