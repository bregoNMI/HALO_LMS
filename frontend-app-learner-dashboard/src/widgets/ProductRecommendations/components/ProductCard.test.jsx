import React from 'react';
import { shallow } from '@edx/react-unit-test-utils';

import { mockCrossProductCourses, mockOpenCourses, mockFallbackOpenCourse } from '../testData';
import { trackProductCardClicked, trackCourseCardClicked } from '../optimizelyExperiment';
import { productCardClicked, discoveryCardClicked } from '../track';
import ProductCard from './ProductCard';
import { courseTypeToProductTypeMap } from '../utils';

jest.mock('../optimizelyExperiment', () => ({
  trackProductCardClicked: jest.fn(),
  trackCourseCardClicked: jest.fn(),
}));

jest.mock('../track', () => ({
  productCardClicked: jest.fn(),
  discoveryCardClicked: jest.fn(),
}));

jest.mock('@edx/frontend-platform/auth', () => ({
  getAuthenticatedUser: jest.fn(() => ({ userId: '1' })),
}));

describe('ProductRecommendations ProductCard', () => {
  const getProps = (course) => {
    const {
      title,
      owners: [{ name: subtitle }],
      image: { src: headerImage },
      owners: [{ logoImageUrl: schoolLogo }],
    } = course;

    return {
      title,
      subtitle,
      headerImage,
      schoolLogo,
      courseType: courseTypeToProductTypeMap[course.courseType],
      courseRunKey: course.courseRunKey,
      url: course.marketingUrl,
    };
  };

  const crossProductProps = getProps(mockCrossProductCourses[0]);
  const openCourseProps = getProps(mockOpenCourses[0]);
  const fallbackOpenCourseProps = getProps(mockFallbackOpenCourse[0]);

  it('matches snapshot', () => {
    expect(shallow(<ProductCard {...crossProductProps} />).snapshot).toMatchSnapshot();
  });

  it('has the query string parameter attached to a fallback recommendations url', () => {
    const wrapper = shallow(<ProductCard {...fallbackOpenCourseProps} />);
    const cardUrl = wrapper.instance.findByType('Card')[0].props.destination;

    expect(cardUrl).toEqual('https://www.edx.org/course/some-course?linked_from=recommender');
  });

  it('send outs experiment events related to open courses when clicked', () => {
    const wrapper = shallow(<ProductCard {...openCourseProps} />);
    const { courseRunKey, title, url } = openCourseProps;

    wrapper.instance.props.onClick();

    expect(trackCourseCardClicked).toHaveBeenCalledWith('1');
    expect(discoveryCardClicked).toHaveBeenCalledWith(courseRunKey, title, `${url}&linked_from=recommender`);
  });

  it('send outs experiment events related to cross product courses when clicked', () => {
    const wrapper = shallow(<ProductCard {...crossProductProps} />);
    const {
      courseRunKey,
      title,
      courseType,
      url,
    } = crossProductProps;

    wrapper.instance.props.onClick();

    expect(trackProductCardClicked).toHaveBeenCalledWith('1');
    expect(productCardClicked).toHaveBeenCalledWith(courseRunKey, title, courseType, `${url}&linked_from=recommender`);
  });
});
