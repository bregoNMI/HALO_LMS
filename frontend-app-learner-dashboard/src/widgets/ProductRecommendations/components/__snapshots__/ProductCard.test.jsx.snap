// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductRecommendations ProductCard matches snapshot 1`] = `
<Card
  as="Hyperlink"
  className="base-card d-flex text-decoration-none"
  destination="https://www.edx.org/course/some-course?utm_source=source&linked_from=recommender"
  isClickable={true}
  onClick={[Function]}
>
  <Card.ImageCap
    logoAlt="logo for Harvard University"
    logoSrc="http://www.image.com/ef72daf3-c9a1-4c00-ba37-b3514392bdcf-8839c516815a.png"
    src="https://www.image-2.com/ed79a49b-64c1-48d2-afdc-054bf921e38d-6a76ceb47dea.small.jpg"
    srcAlt="header image for Introduction to Computer Science"
  />
  <Card.Header
    className="mt-2"
    subtitle={
      <Truncate
        className="product-card-subtitle font-weight-normal"
        lines={1}
      >
        Harvard University
      </Truncate>
    }
    title={
      <Truncate
        className="product-card-title font-weight-bold"
        ellipsis="â€¦"
        lines={3}
      >
        Introduction to Computer Science
      </Truncate>
    }
  />
  <Card.Section>
    <div
      className="product-badge position-absolute"
    >
      <Badge
        className="bg-light-500 text-dark-500"
      >
        Executive Education
      </Badge>
    </div>
  </Card.Section>
</Card>
`;
